src/hooks/useIntersectionObserver/useIntersectionObserver.ts

```diff
@@ -0,0 +1,37 @@
+import { useEffect } from 'react';
+import { unref, type Unreffable } from '../../utils/unref/unref.js';
+
+/**
+ * This hook creates an IntersectionObserver and uses it to observe
+ * the target element.
+ *
+ * For a full explanation of the IntersectionObserver API, check the docs:
+ * https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver
+ *
+ * @param targetOrTargets - The target or targets to observe
+ * @param callback - The callback to fire when the element resizes
+ * @param options - Optional options object
+ */
+
+export function useIntersectionObserver(
```

Comment:
I like the idea of providing multiple targets but people must use memoize, I think it's likely that people will forget this. Maybe we can add some defensive check to warn people about it when they use the hook, if that's not possible we might want to reconsider the option to provide an array of targets.

Reply:
Needed the hook in a project so copied/updated it a bit, this is what I'd do.

```ts
export function useIntersectionObserver(
  target: Unreffable<Element | null>,
  callback: IntersectionObserverCallback,
  options?: IntersectionObserverInit
): void {
  const callbackRef = useRefValue(callback);

  useEffect(() => {
    const intersectionObserverInstance = new IntersectionObserver(
      (entries: Array<IntersectionObserverEntry>, observer: IntersectionObserver) => {
        callbackRef.current?.(entries, observer);
      },
      options
    );

    const element = unref(target);

    if (element) {
      intersectionObserverInstance.observe(element);
    }

    return () => {
      intersectionObserverInstance.disconnect();
    };
  }, [options, refCallback, target]);
}
```